### PR TITLE
lightgun support in udev

### PR DESCRIPTION
**Pull request by hiddenasbestos**
## Description

This is a blind attempt to implement lightgun support in the udev driver. I will be relying on travis to find bugs with this (I'll wager it won't compile first time) and then @twinaphex to test it. Please don't merge unless you at least get a green light and no warnings, I'll probably reply to say I'm happy :)

## Related Issues

Connected to #5827

## Related Pull Requests

No dependencies.

## Reviewers

Here you go @twinaphex  !

### CODE REVIEW

**Comment by hiddenasbestos** [comment]
Erm, looks like two things are happening here: 1. travis isn't building with udev and 2. something weird is broken with the build since it's got errors somewhere I've not gone near.

**Comment by hiddenasbestos** [comment]
Okay so the Travis build has passed but this udev code was not part of the build.

**Comment by twinaphex** [comment]
Hi there, I tested the PR.

The buttons seem to work but X/Y mouse input seems to not work correctly. Target reticule does not appear for me correctly and X/Y input seems to not properly correspond to mouse input.

File: input/drivers/udev_input.c [hunk 1]
```diff
@@ -757,6 +766,49 @@ static bool udev_pointer_is_off_window(const udev_input_t *udev)
...
+   int16_t res_screen_y        = 0;
+
+   vp.x                        = 0;
+   vp.y                        = 0;
+   vp.width                    = 0;
+   vp.height                   = 0;
+   vp.full_width               = 0;
+   vp.full_height              = 0;
+
+   udev_input_mouse_t *mouse = udev_get_mouse(udev, port);
```
Comment (twinaphex) [inline input/drivers/udev_input.c]: Build error. Changed this to -

udev_input_mouse_t *mouse = udev_get_mousee(udev, idx);

File: input/drivers/udev_input.c [hunk 2]
```diff
@@ -877,31 +967,79 @@ static int16_t udev_input_state(void *data,
...
+
       case RETRO_DEVICE_LIGHTGUN:
-         return udev_lightgun_state(udev, port, id);
+         switch ( id )
+         {
+            /*aiming*/
+            case RETRO_DEVICE_ID_LIGHTGUN_SCREEN_X:
+            case RETRO_DEVICE_ID_LIGHTGUN_SCREEN_Y:
+            case RETRO_DEVICE_ID_LIGHTGUN_IS_OFFSCREEN:
+               return udev_lightgun_aiming_state( x11, idx, id );
```
Comment (twinaphex) [inline input/drivers/udev_input.c]: Build error. Change x11 to udev.

File: input/drivers/udev_input.c [hunk 3]
```diff
@@ -877,31 +967,79 @@ static int16_t udev_input_state(void *data,
...
+               return udev_is_pressed(udev, joypad_info, binds[port], port, RARCH_LIGHTGUN_DPAD_DOWN);
+            case RETRO_DEVICE_ID_LIGHTGUN_DPAD_LEFT:
+               return udev_is_pressed(udev, joypad_info, binds[port], port, RARCH_LIGHTGUN_DPAD_LEFT);
+            case RETRO_DEVICE_ID_LIGHTGUN_DPAD_RIGHT:
+               return udev_is_pressed(udev, joypad_info, binds[port], port, RARCH_LIGHTGUN_DPAD_RIGHT);
+
+            /*deprecated*/
+            case RETRO_DEVICE_ID_LIGHTGUN_X:
+               {
+                  udev_input_mouse_t *mouse = udev_get_mouse(udev, port)
```
Comment (twinaphex) [inline input/drivers/udev_input.c]: Add ; to end of line in order to avoid build error.

File: input/drivers/udev_input.c [hunk 4]
```diff
@@ -877,31 +967,79 @@ static int16_t udev_input_state(void *data,
...
+                  return (mouse) ? : udev_mouse_get_x(mouse) : 0;
+               }
+            case RETRO_DEVICE_ID_LIGHTGUN_Y:
+               {
+                  udev_input_mouse_t *mouse = udev_get_mouse(udev, port)
```
Comment (twinaphex) [inline input/drivers/udev_input.c]: Add ; to end of line in order to avoid build error.

**Comment by hiddenasbestos** [comment]
Hi, thanks for the corrections. I've pushed an update which should fix the problems you've reported.

**Comment by twinaphex** [comment]
Hi there, that should take care of the build errors yeah. The X/Y targeting issues will still remain though. X/Y movement seems to not correspond to that of the actual mouse.

**Comment by retro-wertz** [comment]
is the X/Y movement in question related to when using absolute mode(or
whatever it is called now)?

**Comment by hiddenasbestos** [comment]
Hi @twinaphex, I'm not sure why. I did use udev,port on that first amend which differs slightly from what you suggested.  But if it still doesn't work it, my luck might have run out and we'll need a debugger (or at least some printfs)

**Comment by hiddenasbestos** [comment]
My best guess would be investigate the xy position params passed to video_driver_translate_coord_viewport_wrap.

**Comment by hiddenasbestos** [comment]
Okay that should work. 

You may need to change the "mouse index" on the input binding page - I had to set mine to 3 before it would work. I believe that to be a general udev issue, not related to lightguns.

**Comment by orbea** [comment]
@hiddenasbestos This broke the zapper in nestopia as can be easily tested with Duck Hunt. Basically the mouse clicks no not register anymore.
```
Bisect results.
```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
cbfdd7bbd95456958baca345cc198b2664108977
3dc35344983421ef8b6cc190a493a41373c41869
6f116b2edfc841643f42b136c320fca8dde62476
4e79ef9f61eb872d96a8d67d03478df14037baba
c453675d386606871a32fb04af7b1ea3822ebc93
c76545bf6ce87e3b131cf0e406655b1918b5c9de
We cannot bisect more!
```
https://github.com/libretro/RetroArch/commit/cbfdd7bbd95456958baca345cc198b2664108977
https://github.com/libretro/RetroArch/commit/3dc35344983421ef8b6cc190a493a41373c41869
https://github.com/libretro/RetroArch/commit/6f116b2edfc841643f42b136c320fca8dde62476
https://github.com/libretro/RetroArch/commit/4e79ef9f61eb872d96a8d67d03478df14037baba
https://github.com/libretro/RetroArch/commit/c453675d386606871a32fb04af7b1ea3822ebc93
https://github.com/libretro/RetroArch/commit/c76545bf6ce87e3b131cf0e406655b1918b5c9de

Also see this nestopia issue. https://github.com/libretro/nestopia/issues/49

**Comment by retro-wertz** [comment]
the new lightgun api does not have a default config assigned to the buttons. hence no mouse input. 

> @hiddenasbestos This broke the zapper in nestopia as can be easily tested with Duck Hunt. Basically the mouse clicks no not register anymore.
> 
> Bisect results.
> 
> ```
> There are only 'skip'ped commits left to test.
> The first bad commit could be any of:
> cbfdd7bbd95456958baca345cc198b2664108977
> 3dc35344983421ef8b6cc190a493a41373c41869
> 6f116b2edfc841643f42b136c320fca8dde62476
> 4e79ef9f61eb872d96a8d67d03478df14037baba
> c453675d386606871a32fb04af7b1ea3822ebc93
> c76545bf6ce87e3b131cf0e406655b1918b5c9de
> We cannot bisect more!
> ```
> 
> [cbfdd7b](https://github.com/libretro/RetroArch/commit/cbfdd7bbd95456958baca345cc198b2664108977)
> [3dc3534](https://github.com/libretro/RetroArch/commit/3dc35344983421ef8b6cc190a493a41373c41869)
> [6f116b2](https://github.com/libretro/RetroArch/commit/6f116b2edfc841643f42b136c320fca8dde62476)
> [4e79ef9](https://github.com/libretro/RetroArch/commit/4e79ef9f61eb872d96a8d67d03478df14037baba)
> [c453675](https://github.com/libretro/RetroArch/commit/c453675d386606871a32fb04af7b1ea3822ebc93)
> [c76545b](https://github.com/libretro/RetroArch/commit/c76545bf6ce87e3b131cf0e406655b1918b5c9de)
> 
> Also see this nestopia issue. [libretro/nestopia#49](https://github.com/libretro/nestopia/issues/49)

**Comment by orbea** [comment]
@retro-wertz So what are you saying the fix is? Breaking cores is not good...

**Comment by orbea** [comment]
This also affects fceumm.

**Pull request merged**